### DISCUSSION TITLE
Using Freeplane as a bookmark manager for audio and video files.

### DISCUSSION THREAD

**Discussion by aaa1386** [freeplane/freeplane#2166] [Ideas]
<p><strong>Imagine Freeplane evolving into a powerful tool that not only organizes your thoughts but also efficiently manages audio and video bookmarks.</strong> Think about how convenient it would be to navigate large media files directly from your mind map, jumping to specific points in a file with a single click. This enhancement could revolutionize how users work with multimedia content, saving time and providing unparalleled organization.</p>

<h3>The Challenge:</h3>
<p>Many media players (e.g., <strong>PotPlayer</strong>) support adding bookmarks to audio and video files, allowing users to mark specific timestamps for easy navigation. For example, in the image below, I've created three bookmarks ("AAA", "BBB", "CCC") in an audio file at specific timestamps. While bookmarks are very useful, managing a large number of bookmarks across multiple media files can become challenging—especially when users need a structured way to categorize and access them efficiently.</p>

![image](https://github.com/user-attachments/assets/aa89c774-181c-41f7-ac35-b82408fa488b)

<p><strong>Currently, Freeplane allows users to link audio and video files to nodes.</strong> When clicked, the linked file plays from the beginning. While this is a great feature, there’s an untapped opportunity to make it even more powerful by integrating bookmark management into Freeplane’s functionality.</p>

<h3>Current Functionality:</h3>
<p><strong>Freeplane allows linking audio and video files to nodes.</strong> When the user clicks on the link, the associated file is played. This is a great feature for organizing media content within mind maps.</p>

<h3>Suggested Solution:</h3>
<p><strong>Freeplane could enhance its media link functionality by allowing users to specify a start time for linked audio or video files.</strong></p>

<ul>
  <li><strong>Define a Start Time for Linked Media:</strong>
    <ul>
      <li>When linking an audio or video file to a node, users should have the option to specify a start time for playback.</li>
      <li>For example, when linking an audio file, Freeplane could prompt:
        <br>"Do you want to start playback from the beginning or a specific time?"
      </li>
      <li>If the user selects a specific time, they can enter the desired time (e.g., 00:02:10 for 2 minutes and 10 seconds).</li>
      <li>The node text can then serve as a title or label for bookmarks previously created in the audio file.</li>
    </ul>
  </li>
</ul>

<ul>
  <li><strong>Organize Bookmarked Media in Freeplane:</strong>
    <ul>
      <li>Users can create and manage nodes with linked bookmarks, categorizing them into  groups based on themes, projects, or any custom hierarchy.</li>
    </ul>
  </li>
</ul>

<h3>Why Freeplane?</h3>
<p><strong>While it is possible to play a specific time from an audio file using small .bat files with two lines of code, these solutions lack the organization and flexibility of Freeplane.</strong></p>
<p><strong>Freeplane’s node structure and categorization capabilities make it an ideal tool for managing bookmarks across multiple audio and video files.</strong></p>

<h3>Benefits:</h3>
<ul>
  <li><strong>Streamlined Bookmark Management:</strong> Users can organize bookmarks for multiple files in a centralized, visual way.</li>
  <li><strong>Improved Workflow:</strong> Instead of manually searching through bookmarks in media players, users can directly jump to specific points in media files from Freeplane.</li>
  <li><strong>Flexibility:</strong> Users can categorize bookmarks in custom hierarchies, aligning with their workflow or projects.</li>
</ul>

<h3>Example Workflow:</h3>
<ol>
  <li>The user links an audio file to a node in Freeplane.</li>
  <li>During the linking process, Freeplane asks if the user wants to start playback from the beginning or a specific time.</li>
  <li>The user enters the time (e.g., 00:02:10).</li>
  <li>The node now serves as a bookmark, with its text serving as the bookmark label.</li>
  <li>The user can organize multiple such nodes into categories for easy access.</li>
</ol>

The image below is the same three bookmarks that you saw in the image above. Here is two simple categorization of them in FP.
 
![image](https://github.com/user-attachments/assets/dacf067c-98fb-4964-bad6-f32399dfacdb))

<h3>Conclusion</h3>
<p><strong>By allowing users to specify start times for media playback, Freeplane can become a powerful tool for managing bookmarks across audio and video files.</strong> This simple yet impactful enhancement would open new possibilities for users, making Freeplane an even more versatile and indispensable tool in their workflows.</p>

**Comment by euu2021** [comment]
There is already a script to navigate to specific pages in PDFs: https://sourceforge.net/p/freeplane/discussion/758437/thread/e8f385c604/

In order to make it work in other media bookmarks, it's necessary to find a player that accepts a command line specifying a bookmark.

I see that PotPlayer has this command:
`/seek=time :Starts playback of the specified/last played content from the specified time point.`

So, maybe it's possible to extract the timestamp of the bookmark and put it in the command. For that, maybe, it's easier to save the bookmarks directly in FP. You can create a shortcut to easily save timestamps:
![image](https://github.com/user-attachments/assets/9c195302-5aa8-4697-9fa6-647888298dec)

Does that sound viable, or is there a better way?

**Comment by nnako** [comment]
Hi,

as @euu2021 wrote, you need an application that is able to be called from the command line and accepts respective timing information. Please, see the script below which is able to play any MP4 media file using  the VLC media player. The only prerequisites are:

- install the appropriate media player application (e.g. VLC media player)
- insert the correct application path (here: VLC) into the respective variable `path_VlcApp` in the code
- provide the absolute media file link within any node's link portion
- provide the timing information in a fix format (like `34m45s`) in one of its child nodes core
- select the node holding the time
- start the script (I'd recommend to set its execution on a specific hotkey)

Whenever you select a node holding a time value, the script will look for a parent node holding the desired media file path and then call the VLC application, transfers the target media file path as well as the time information to it. VLC then (hopefully) jumps to the time within the given media file and plays the MP4. It also works for MP3 audio files.

Here an example within a mindmap holding different times and subjects:
![image](https://github.com/user-attachments/assets/1d7f143f-1c96-4547-9080-8a468d55e249)

The `link` portion of the green node must hold an absolute path to the desired media file, like:
`file:/d:\PROJEKTE\_work/MediaFile.mp4`

So, just select the node holding the time value and start the script.

Here the script:

```groovy
// @ExecutionModes({ON_SINGLE_NODE="/menu_bar/nnako"})




//
// set constants and init values
//

// application paths
def path_VlcApp                         = "C:\\Program Files\\VideoLAN\\VLC\\vlc.exe"




//
// CHECK for VIDEO or AUDIO time spec
//

// When the current node contains a hyperlink into a VIDEO file or a URL, it is assumed
// that the user links from his mindmap into the start of a video file. If the
// current node carries characteristic start time values like "1h30m45s", it is
// assumed that the node itself or some node on the parent path specifies the
// URL or file location to the video.
//
//  .-------------------.                           .--------.
//  | <text> / <refref> +- ... some more nodes ... -+ <time> |
//  '-------------+-----'                           '--------'
//                |
//                '- <link>
// 
//  .-----------.
//  | <reftext> +                node containing an actual link (somewhere in the map)
//  '------+----'
//         |
//         '- <link>
//
//
// OPTIONS
// 
// <time>                                e.g. 1h30m23s
// <core>                                e.g. 1h30m23s
// <link> = [<prot>:/][<drv>:/]<path>[#<time>]    e.g. mp4:/c:/temp/videodocument.mp4#1h29m23s

// check for specific file link marker or specific file extension and a
// specific meta information token within the link portion of the current node.
if(
        (
          ! node.link.text.equals(null)
            // check for local video file
            &&
              (
                (
                  (
                    node.link.text.toLowerCase().startsWith( 'video:/' )
                    || node.link.text.toLowerCase().startsWith( 'audio:/' )
                    || node.link.text.toLowerCase().startsWith( 'mp4:/' )
                    || node.link.text.toLowerCase().indexOf(".mp4") > 0
                    || node.link.text.toLowerCase().indexOf(".mp3") > 0
                  )
                  // only if parameter token present. otherwise a standard
                  // <CTRL> + <ENTER> should open the file directly
                  && node.link.text.toLowerCase().indexOf("#") > 0
                )
                || 
                (
                  node.link.text.toLowerCase().startsWith( 'https:/' )
                )
                  // only if time parameter token present. otherwise a
                  // standard <CTRL> + <ENTER> should open the file directly
                  && node.link.text.toLowerCase().indexOf("t=") > 0
              )
            )
            ||
            (
              check_for_timestamp(node.text)
            )
        ){




    //
    // get time parameter from current node
    //

    // decode timestamp from core
    def strTime = node.text
    def iTimeInSeconds = decode_timestamp_into_seconds(node.text)




    //
    // get video or audio path
    //

    // look for video path in current node and possibly its parents if core
    // contains timestamp
    def check = node
    def max_levels = 4
    def bVideoPathFound = false
    for(i=0; i<max_levels; i++){ 
        // IF no direct video link path within current node
        if( check_for_video_or_audio_path(check.link.text) ){
            bVideoPathFound = true
            break
        }
        // IF no direct video link path within linked node
        else if(
                ! check.link.text.equals(null)
                &&
                check.link.text.startsWith("#") 
                &&
                check_for_video_or_audio_path(c.find{it.id==check.link.text[1..check.link.text.length()-1]}[0].link.text)
        ){
            bVideoPathFound = true
            check = c.find{it.id==check.link.text[1..check.link.text.length()-1]}[0]
            break
        }
        // ELSE move one node further up
        else {
            check = check.parent
        }
    }
    if( bVideoPathFound ){

        // store path
        def path_UserFile = check.link.text




        //
        // evaluate CLI parameters
        //
        def iSeconds = -1

        // check for WebLink (not local file)
        bWebLink = check_for_weblink(path_UserFile.toLowerCase())

        // cut off leading protocol
        if( bWebLink == false ){
            path_UserFile = remove_protocol_info_from_abs_path(path_UserFile)
        }

        // create absolute document path
        if( bWebLink == false && path_UserFile[1] != ":" ){
            path_UserFile = node.mindMap.file.parent + "\\" + path_UserFile
        }

        // if the path contains a parameter
        if( bWebLink == false && path_UserFile.indexOf("#") >= 0 ){
            // get time stamp
            iSeconds = decode_timestamp_into_seconds(
                    path_UserFile.substring(path_UserFile.indexOf("#")+1)
                    )
            // cut down file path
            path_UserFile = path_UserFile.substring(0, path_UserFile.indexOf("#"))
        }




        //
        // sanitize path string
        //

        if( bWebLink == false ){

            // replace slash with backslash
            path_UserFile = path_UserFile.replaceAll( "/", "\\\\" )

            // replace space representation with actual space
            path_UserFile = path_UserFile.replaceAll( "%20", " " )

            // remove mysterious "smb:" if present
            path_UserFile = path_UserFile.replaceAll( "smb:", "" )
        }

        // time stamp
        def strTimeInSeconds = ""
        if( iTimeInSeconds >= 0 ){
            // override with value from core
            strTimeInSeconds = iTimeInSeconds.toString()

        } else if( iSeconds >= 0 ){
            strTimeInSeconds = iSeconds.toString()
        }




        //
        // execute application on CLI
        //

        if( bWebLink == false ){
            if( strTimeInSeconds.equals("") ){
                def proc = "\"$path_VlcApp\" \"$path_UserFile\"".execute()
                c.statusInfo = "opened video file in external player"
            } else {
                def proc = "\"$path_VlcApp\" --start-time $strTimeInSeconds \"$path_UserFile\"".execute()
                c.statusInfo = "opened video file in external player at StartTime: " + strTimeInSeconds
            }
        } else {
            showPageInBrowser( path_UserFile + "&t=" + strTime )
        }

    }


}




//
// GENERAL FUNCTIONS
//




/**
 *  remove possible leading protocol token from path string. this function
 *  works only for the Windows file system as here, absolute paths start with
 *  single drive letters.
 *
 *  ...
 */
private String remove_protocol_info_from_abs_path( strUrl ){

    def pos = strUrl.indexOf(":/")
    def strPath = strUrl
    if( pos>2 ){
        // cut off protocol token
        strPath = strPath.substring(pos+2)
        // cut off any further leading slash
        pos = strPath.indexOf("/")
        while( pos==0 ){
            strPath = strPath.substring(1)
            pos = strPath.indexOf("/")
        }
    }
    return strPath
}




/**
 *  decode timestamp given as string into integer value. the timestamp has to
 *  follow the format
 *
 *  [[<hours>h]<minutes>m]<seconds>s
 */
private Integer decode_timestamp_into_seconds( strTime ){

    // check for time information within node core
    def match = strTime.toLowerCase() =~ /^(\d+h)?(\d+m)?(\d+s)/ // REGEX_TIMESTAMP

    // decode start time definition
    def iSeconds = 0
    if( match ){

        // hours
        if( match[0][1] ){
            iSeconds += 60 * 60 * match[0][1].substring(0, match[0][1].length()-1).toInteger()
        }
        // minutes
        if( match[0][2] ){
            iSeconds += 60 * match[0][2].substring(0, match[0][2].length()-1).toInteger()
        }
        // seconds
        if( match[0][3] ){
            iSeconds += match[0][3].substring(0, match[0][3].length()-1).toInteger()
        }

        return iSeconds
    }
    return -1
}




/**
 *  check if condition is met for exaluating a time stamp
 */
private boolean check_for_timestamp( strCheck ){
    def match = strCheck.toLowerCase() =~ /^(\d+h)?(\d+m)?(\d+s)/ // REGEX_TIMESTAMP
    if( match ){
        return true
    }
    return false
}


/**
 *  check if string contains a video path
 */
private boolean check_for_video_or_audio_path( strCheck ){
    if(
            ! strCheck.equals(null)
            && (
                strCheck.toLowerCase().startsWith( 'video:/' )
                || strCheck.toLowerCase().startsWith( 'audio:/' )
                || strCheck.toLowerCase().startsWith( 'https:/' )
                || strCheck.toLowerCase().startsWith( 'mp4:/' )
                || strCheck.toLowerCase().indexOf(".mp4") > 0
                || strCheck.toLowerCase().indexOf(".mp3") > 0
                )
            //&& strCheck.toLowerCase().indexOf("#") > 0
            ){
        return true
    }
    c.statusInfo = strCheck + "!"
    return false
}


/**
 *  check if url is a web link
 */
private boolean check_for_weblink( strCheck ){
    if(
            ! strCheck.equals(null)
            && strCheck.toLowerCase().startsWith( 'https:/' )
            ){
        return true
    }
    c.statusInfo = strCheck + "!"
    return false
}
```

**Reply by aaa1386** [reply]
Thank you. Can you configure the script to work with PotPlayer?

**Reply by nnako** [reply]
Hi @aaa1386 ,

sorry for the late reply. In order to adapt the script to your own choice for a media player, you should be fine after doing those steps:

- install your favorite media player (in your case PotPlayer)
- identify the location of that executable on your system (it could be something like C:\programs\PotPlayer\player.exe)
- edit the first non-comment code line within the script to hold that location:

```groovy
def path_VlcApp                         = "C:\\Program Files\\VideoLAN\\VLC\\vlc.exe"
```

- now, check in which way your media player expects parameters on the command line
- for VLC, the parameters must be e.g. `--start-time 1000`, where `1000` is the start time in seconds, then followed by the path to the media file
- within the script's section "execute application on CLI" you will see how it was done (Line#10):

```groovy
        //
        // execute application on CLI
        //

        if( bWebLink == false ){
            if( strTimeInSeconds.equals("") ){
                def proc = "\"$path_VlcApp\" \"$path_UserFile\"".execute()
                c.statusInfo = "opened video file in external player"
            } else {
                def proc = "\"$path_VlcApp\" --start-time $strTimeInSeconds \"$path_UserFile\"".execute()
                c.statusInfo = "opened video file in external player at StartTime: " + strTimeInSeconds
            }
        } else {
            showPageInBrowser( path_UserFile + "&t=" + strTime )
        }
```

- when PotPlayer expects other parameters on the command line, they would have to be defined, here

Please tell me if you had success.

**Reply by aaa1386** [reply]
> * install your favorite media player (in your case PotPlayer)

 Hi .

Did I do it right?


![image](https://github.com/user-attachments/assets/5d7e7370-b123-4604-aa85-d0e16bbc899d)

![image](https://github.com/user-attachments/assets/b79c9d13-ac5d-411f-8f2d-8837337602ed)

https://github.com/user-attachments/assets/ef90be85-2d6f-4c99-940e-f5f4b776e3dc

**Comment by Lotus49b** [comment]
First let me say that I'm a big fan of Freeplane. I use it a lot, and it's among the 3 most used apps on my laptop (above MS Ofiice even!).

But I've observed this phenomenon quite a lot in some forums, where people who love an app try to "bend" it to a purpose that it's simply not designed for. This feature request suggests that again.

My most used app is Obsidian, a Personal Knowledge Management system. It's very extensible through the use of a large plug-ins library, and I use it (among other things) for annotating and summarising videos. You can use an AI plugin to analyse the video and create bookarks from the transcript, or just do that yourself in a window while the video is playing. You can embed the video in your summary note, or simply link from bookmarks in your note to the actual video in it's original place. 

The thing about this is that it's a much faster and more intuitive workflow than the one you are suggesting, you don't have to code to get a result, and the bookmarks and notes are much more useful through Obsidian's ability to search bookmarks, tags and links to other notes with related information.

I suggest you look at the Obisidian website and also the user forums for more information.

One more thing - Obsidian users often ask for better mind mapping functionality in that app and then I point them to Freeplane. As the saying goes, when the only tool in your toolbox is a hammer then everything looks like a nail.

**Reply by euu2021** [reply]
I like the tip, thanks.

> But I've observed this phenomenon quite a lot in some forums, where people who love an app try to "bend" it to a purpose that it's simply not designed for.

All those things are possible on Obsidian because people "bent" it by creating the plugins. The same could be done in Freeplane, as it's extremely customizable by scripts. 

For me, the advantage of doing everything inside Freeplane is that I can concentrate all my knowledge management there, in a way that I can link all the information easily.

**Reply by lilive** [reply]
Hi,
I have something to share here :slightly_smiling_face:, but this is not a "ready to go" solution for you, @ajaberi1360 :worried: 

I have a global keyboard shortcut that allows me to create various Freeplane nodes links. With this shortcut, I can bookmark :
- the file selected in my file browser
- the webpage I'm reading with Firefox
- the pdf page I'm reading with SumatraPDF
- the video file I'm watching with [mpv](https://mpv.io/), with its current playback position 
- the video I'm watching in youtube with Firefox, with its current playback position 
- the Info page I'm reading in Emacs

The usecase is:
I have a Freeplane map opened. I'm using one of the above application. I press the keyboard shortcut. A node in Freeplane is created as a sibling of the currently selected node. It links to the content I was reading / watching. Later, I can open the same content, in the same application, at the same page / playback position, with the Freeplane "Follow link" feature.

This system is extensible, and may be used with other applications, at the cost of tweaking and scripting. Once setup, I find it really handy.
It relies on a single Freeplane script that create the bookmark node from the data in the clipboard, and on AHK scripts that copy the proper data in the clipboard. So, this is Windows only, sorry. Some of these AHK scripts rely themselves on other scripts. For example, I had to code a little lua script for mpv, to copy the playback time. Same thing for Firefox, I have a little addon that I use to copy the webpage title, or the youtube video title and its playback time.

I add the files to this post. For more information, read the comments in the scripts, starting with `AHK scripts > Create Freeplane link.ahk`
[share.zip](https://github.com/user-attachments/files/18267173/share.zip)

**Comment by aaa1386** [comment]
Thanks everyone
I was able to achieve what I wanted with this script

<img width="778" height="273" alt="image" src="https://github.com/user-attachments/assets/16144c6a-f55e-4451-a2e6-e201a2d351ab" />

**Reply by aaa1386** [reply]
```
import java.net.URLDecoder

def timeNode = node
def parentNode = timeNode.parent

def link = parentNode.link?.text ?: ""
if (!link) {
    ui.errorMessage("لینک فایل در گره پدر وجود ندارد.")
    return
}

def mediaPath = link.replaceFirst(/^file:\/+/, "")
mediaPath = URLDecoder.decode(mediaPath, "UTF-8")

def file = new File(mediaPath)
if (!file.exists()) {
    ui.errorMessage("فایل یافت نشد:\n" + mediaPath)
    return
}

def timeText = timeNode.text.trim()
def totalSeconds = 0

if (timeText =~ /^\d{1,2}m\d{1,2}s$/) {
    def matcher = (timeText =~ /(\d+)m(\d+)s/)
    if (matcher.find()) {
        totalSeconds = matcher.group(1).toInteger() * 60 + matcher.group(2).toInteger()
    }
} else if (timeText =~ /^\d{1,2}:\d{2}$/) {
    def parts = timeText.split(":")
    totalSeconds = parts[0].toInteger() * 60 + parts[1].toInteger()
} else {
    ui.errorMessage("زمان باید به شکل 2m30s یا 02:30 وارد شود.")
    return
}

def potplayerPath = "C:\\Program Files\\DAUM\\PotPlayer\\PotPlayerMini64.exe"

def command = [
    "cmd", "/c",
    "start", "",
    "\"${potplayerPath}\"",
    "\"${mediaPath}\"",
    "/seek=${totalSeconds}"
]

def process = command.execute()
process.waitFor()


```
